File: takehome/app/src/main/java/com/fivestars/takehome/CordovaService.java
package com.fivestars.takehome;

import android.app.Service;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.webkit.WebViewClient;
import android.widget.FrameLayout;

import org.apache.cordova.Config;
import org.apache.cordova.ConfigXmlParser;
import org.apache.cordova.CordovaInterfaceImpl;
import org.apache.cordova.CordovaPreferences;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CordovaWebViewEngine;
import org.apache.cordova.CordovaWebViewImpl;
import org.apache.cordova.PluginEntry;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

abstract public class CordovaService extends Service {

    protected WindowManager mWindowManager;
    protected View contentView;
    protected CordovaWebView appView;
    protected CordovaPreferences preferences;
    protected CordovaInterfaceImpl cordovaInterface;
    protected String launchUrl;
    protected ArrayList<PluginEntry> pluginEntries;
    protected WindowManager.LayoutParams params;

    @Override
    public void onCreate() {
        super.onCreate();
        loadConfig();
        cordovaInterface = makeCordovaInterface();
        loadUrl(launchUrl);

        //Inflate the chat head layout we created
        contentView = LayoutInflater.from(this).inflate(getLayoutToInflate(), nullgit status;

        appView.getView().setLayoutParams(new ViewGroup.LayoutParams(dpToPx(180), dpToPx(180)));

        ((ViewGroup) contentView.findViewById(getWebViewParentLayoutId())).addView(appView.getView());

        //Add the view to the window.

        params = new WindowManager.LayoutParams(
                dpToPx(192),
                dpToPx(192),
                WindowManager.LayoutParams.TYPE_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT);

        //Add the view to the window
        mWindowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        mWindowManager.addView(contentView, params);
    }

    public void loadUrl(String url) {
        if (appView == null) {
            init();
        }
        appView.loadUrlIntoView(url, true);
    }

    protected void loadConfig() {
        ConfigXmlParser parser = new ConfigXmlParser();
        parser.parse(this);
        preferences = parser.getPreferences();
        launchUrl = parser.getLaunchUrl();
        pluginEntries = parser.getPluginEntries();
        Config.parser = parser;
    }

    protected void init() {
        appView = makeWebView();
        createViews();
        if (!appView.isInitialized()) {
            appView.init(cordovaInterface, pluginEntries, preferences);
        }
        cordovaInterface.onCordovaInit(appView.getPluginManager());
    }

    //Suppressing warnings in AndroidStudio
    @SuppressWarnings({"deprecation", "ResourceType"})
    protected void createViews() {
        //Why are we setting a constant as the ID? This should be investigated
        appView.getView().setId(100);
        appView.getView().setLayoutParams(new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));

//        setContentView(appView.getView());

        if (preferences.contains("BackgroundColor")) {
            try {
                int backgroundColor = preferences.getInteger("BackgroundColor", Color.BLACK);
                // Background of activity:
                appView.getView().setBackgroundColor(backgroundColor);
            }
            catch (NumberFormatException e){
                e.printStackTrace();
            }
        }

        appView.getView().requestFocusFromTouch();
    }

    /**
     * Construct the default web view object.
     * <p/>
     * Override this to customize the webview that is used.
     */
    protected CordovaWebView makeWebView() {
        return new CordovaWebViewImpl(makeWebViewEngine());
    }

    protected CordovaWebViewEngine makeWebViewEngine() {
        return CordovaWebViewImpl.createEngine(this, preferences);
    }

    protected CordovaInterfaceImpl makeCordovaInterface() {
        return new CordovaInterfaceImpl(this) {
            @Override
            public Object onMessage(String id, Object data) {
                // Plumb this to CordovaActivity.onMessage for backwards compatibility
                return CordovaService.this.onMessage(id, data);
            }
        };
    }

    /**
     * Called when a message is sent to plugin.
     *
     * @param id   The message id
     * @param data The message data
     * @return Object or null
     */
    public Object onMessage(String id, Object data) {
        if ("onReceivedError".equals(id)) {
            JSONObject d = (JSONObject) data;
            try {
                this.onReceivedError(d.getInt("errorCode"), d.getString("description"), d.getString("url"));
            } catch (JSONException e) {
                e.printStackTrace();
            }
        } else if ("exit".equals(id)) {
            stopSelf();
        }
        return null;
    }

    /**
     * Report an error to the host application. These errors are unrecoverable (i.e. the main resource is unavailable).
     * The errorCode parameter corresponds to one of the ERROR_* constants.
     *
     * @param errorCode   The error code corresponding to an ERROR_* value.
     * @param description A String describing the error.
     * @param failingUrl  The url that failed to load.
     */
    public void onReceivedError(final int errorCode, final String description, final String failingUrl) {
        final CordovaService me = this;

        Handler handler = new Handler(Looper.getMainLooper());

        // If errorUrl specified, then load it
        final String errorUrl = preferences.getString("errorUrl", null);
        if ((errorUrl != null) && (!failingUrl.equals(errorUrl)) && (appView != null)) {
            // Load URL on UI thread
            handler.post(new Runnable() {
                public void run() {
                    me.appView.showWebPage(errorUrl, false, true, null);
                }
            });
        }
        // If not, then display error dialog
        else {
            final boolean exit = !(errorCode == WebViewClient.ERROR_HOST_LOOKUP);
            handler.post(new Runnable() {
                public void run() {
                    if (exit) {
                        me.appView.getView().setVisibility(View.GONE);
                    }
                }
            });
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (contentView != null) mWindowManager.removeView(contentView);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private int dpToPx(int dp) {
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
        return Math.round(dp * (displayMetrics.xdpi / DisplayMetrics.DENSITY_DEFAULT));
    }

    abstract int getLayoutToInflate();

    abstract int getWebViewParentLayoutId();
}

File: takehome/app/src/main/java/com/fivestars/takehome/CordovaService.kt
package com.fivestars.takehome

import android.app.Service
import android.content.Intent
import android.graphics.PixelFormat
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.webkit.WebViewClient
import com.fivestars.communication.CommunicationPlugin
import org.apache.cordova.*

import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList

abstract class CordovaService : Service() {

    protected var mWindowManager: WindowManager? = null
    protected var contentView: View? = null
    private var appView: CordovaWebView? = null
    private var preferences: CordovaPreferences? = null
    private var cordovaInterface: CordovaInterfaceImpl? = null
    private var launchUrl: String? = null
    private var pluginEntries: ArrayList<PluginEntry>? = null
    protected var params: WindowManager.LayoutParams? = null

    internal abstract val layoutToInflate: Int

    internal abstract val appViewParentLayoutId: Int

    override fun onCreate() {
        super.onCreate()
        cordovaInterface = makeCordovaInterface()
        loadConfig()
        launchUrl?.let{
            loadUrl(it)
        }

        //Inflate the layout to show to the user
        contentView = LayoutInflater.from(this).inflate(layoutToInflate, null)

        (contentView?.findViewById<View>(appViewParentLayoutId) as ViewGroup).addView(appView?.view)

        //Add the view to the window.

        val viewSize = applicationContext.resources.getDimension(R.dimen.five_star_size).toInt()

        params = WindowManager.LayoutParams(
                viewSize,
                viewSize,
                WindowManager.LayoutParams.TYPE_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT)

        //Add the view to the window
        mWindowManager = getSystemService(WINDOW_SERVICE) as WindowManager
        mWindowManager?.addView(contentView, params)

        var communicationPlugin = appView?.pluginManager?.getPlugin("CommunicationPlugin") as CommunicationPlugin

        val handler = Handler(Looper.getMainLooper())

        var count = 777

        var runThis = object : Runnable {
            override fun run() {
                communicationPlugin.callbackContext?.let{
                    Log.e("darran", "we have a callback context")
                    var result =  PluginResult(PluginResult.Status.OK, communicationPlugin.getTransactionResponsePayload(count).toString())
                    result.keepCallback = true
                    it.sendPluginResult(result)
                    count += 777
                }
                handler.postDelayed(this as Runnable, 1000)
            }
        }

        handler.post(runThis)
    }

    private fun loadUrl(url: String) {
        if (appView == null) {
            init()
        }
        appView?.loadUrlIntoView(url, true)
    }

    private fun loadConfig() {
        val parser = ConfigXmlParser()
        parser.parse(this)
        preferences = parser.preferences
        launchUrl = parser.launchUrl
        pluginEntries = parser.pluginEntries
        Config.parser = parser
    }

    private fun init() {
        appView = makeWebView()
        createViews()
        if (!appView!!.isInitialized) {
            appView!!.init(cordovaInterface, pluginEntries, preferences)
        }
        cordovaInterface?.onCordovaInit(appView!!.pluginManager)
    }

    //Suppressing warnings in AndroidStudio
    private fun createViews() {
        appView?.view?.requestFocusFromTouch()
    }

    /**
     * Construct the default web view object.
     *
     *
     * Override this to customize the webview that is used.
     */
    private fun makeWebView(): CordovaWebView {
        return CordovaWebViewImpl(makeWebViewEngine())
    }

    private fun makeWebViewEngine(): CordovaWebViewEngine {
        return CordovaWebViewImpl.createEngine(this, preferences)
    }

    private fun makeCordovaInterface(): CordovaInterfaceImpl {
        return object : CordovaInterfaceImpl(this) {
            override fun onMessage(id: String, data: Any): Any? {
                // Plumb this to CordovaActivity.onMessage for backwards compatibility
                return this@CordovaService.onMessage(id, data)
            }
        }
    }

    /**
     * Called when a message is sent to plugin.
     *
     * @param id   The message id
     * @param data The message data
     * @return Object or null
     */
    fun onMessage(id: String, data: Any): Any? {
        if ("onReceivedError" == id) {
            val d = data as JSONObject
            try {
                this.onReceivedError(d.getInt("errorCode"), d.getString("description"), d.getString("url"))
            } catch (e: JSONException) {
                e.printStackTrace()
            }

        } else if ("exit" == id) {
            stopSelf()
        }
        return null
    }

    /**
     * Report an error to the host application. These errors are unrecoverable (i.e. the main resource is unavailable).
     * The errorCode parameter corresponds to one of the ERROR_* constants.
     *
     * @param errorCode   The error code corresponding to an ERROR_* value.
     * @param description A String describing the error.
     * @param failingUrl  The url that failed to load.
     */
    private fun onReceivedError(errorCode: Int, description: String, failingUrl: String) {
        val me = this

        val handler = Handler(Looper.getMainLooper())

        // If errorUrl specified, then load it
        val errorUrl = preferences?.getString("errorUrl", null)
        if (errorUrl != null && failingUrl != errorUrl && appView != null) {
            // Load URL on UI thread
            handler.post { me.appView?.showWebPage(errorUrl, false, true, null) }
        } else {
            val exit = errorCode != WebViewClient.ERROR_HOST_LOOKUP
            handler.post {
                if (exit) {
                    me.appView?.view?.visibility = View.GONE
                }
            }
        }// If not, then display error dialog
    }

    override fun onDestroy() {
        super.onDestroy()
        if (contentView != null) mWindowManager?.removeView(contentView)
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }
}
